// @chatkitty/core/src/lib/thread.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReadThreadSucceededResult = exports.RetrieveThreadMessageSucceededResult = exports.RetrieveThreadChannelSucceededResult = exports.ListThreadsSucceededResult = exports.CreatedThreadResult = void 0;
const result_1 = require("./result");
class CreatedThreadResult extends result_1.ChatKittySucceededResult {
    constructor(thread) {
        super();
        this.thread = thread;
    }
}
exports.CreatedThreadResult = CreatedThreadResult;
class ListThreadsSucceededResult extends result_1.ChatKittySucceededResult {
    constructor(paginator) {
        super();
        this.paginator = paginator;
    }
}
exports.ListThreadsSucceededResult = ListThreadsSucceededResult;
class RetrieveThreadChannelSucceededResult extends result_1.ChatKittySucceededResult {
    constructor(channel) {
        super();
        this.channel = channel;
    }
}
exports.RetrieveThreadChannelSucceededResult = RetrieveThreadChannelSucceededResult;
class RetrieveThreadMessageSucceededResult extends result_1.ChatKittySucceededResult {
    constructor(message) {
        super();
        this.message = message;
    }
}
exports.RetrieveThreadMessageSucceededResult = RetrieveThreadMessageSucceededResult;
class ReadThreadSucceededResult extends result_1.ChatKittySucceededResult {
    constructor(thread) {
        super();
        this.thread = thread;
    }
}
exports.ReadThreadSucceededResult = ReadThreadSucceededResult;
