// @chatkitty/core/src/lib/user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockUserSucceededResult = exports.CheckUserIsChannelMemberSucceededResult = exports.ListUserSucceededResult = exports.ListUsersSucceededResult = void 0;
const result_1 = require("./result");
class ListUsersSucceededResult extends result_1.ChatKittySucceededResult {
    constructor(paginator) {
        super();
        this.paginator = paginator;
    }
}
exports.ListUsersSucceededResult = ListUsersSucceededResult;
class ListUserSucceededResult extends result_1.ChatKittySucceededResult {
    constructor(user) {
        super();
        this.user = user;
    }
}
exports.ListUserSucceededResult = ListUserSucceededResult;
class CheckUserIsChannelMemberSucceededResult extends result_1.ChatKittySucceededResult {
    constructor(isChannelMember) {
        super();
        this.isChannelMember = isChannelMember;
    }
}
exports.CheckUserIsChannelMemberSucceededResult = CheckUserIsChannelMemberSucceededResult;
class BlockUserSucceededResult extends result_1.ChatKittySucceededResult {
    constructor(user) {
        super();
        this.user = user;
    }
}
exports.BlockUserSucceededResult = BlockUserSucceededResult;
